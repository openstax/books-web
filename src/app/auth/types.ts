export interface User {
  firstName: string;
}

export interface State {
  user: User | undefined;
  established: boolean;
}
