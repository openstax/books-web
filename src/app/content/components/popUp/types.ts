import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
