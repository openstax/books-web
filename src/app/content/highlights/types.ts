import { SerializedHighlight } from '@openstax/highlighter';

export type HighlightData = SerializedHighlight['data'] & {
  note?: string;
};

export interface State {
  enabled: boolean;
  focused?: string;
  highlights: HighlightData[];
}
