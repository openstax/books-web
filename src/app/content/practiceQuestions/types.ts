import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
