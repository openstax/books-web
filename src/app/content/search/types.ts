import { SearchResult } from '@openstax/open-search-client';

export interface State {
  results: SearchResult | null;
  loading: boolean;
  query: null | string;
}
