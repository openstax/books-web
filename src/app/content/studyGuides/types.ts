import { HighlightsSummary } from '@openstax/highlighter/dist/api';

export interface State {
  isEnabled: boolean;
  summary: HighlightsSummary | null;
}
