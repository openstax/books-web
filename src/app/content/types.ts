import { RouteParams, RouteState } from '../navigation/types';
import { content } from './routes';
import { State as SearchState } from './search/types';

export interface Params {
  book: string;
  page: string;
}

export interface State {
  tocOpen: boolean | null;
  params?: Params;
  loading: {
    book?: string;
    page?: string;
  };
  search: SearchState;
  book?: Book;
  page?: Page;
  references: PageReferenceMap[];
}

export interface PageReferenceMap extends PageReference {
  match: string;
}

export interface PageReference {
  state: RouteState<typeof content>;
  params: RouteParams<typeof content>;
}

export interface Book {
  id: string;
  shortId: string;
  title: string;
  theme: 'blue' | 'green' | 'gray' | 'yellow' | 'deep-green' | 'light-blue' | 'orange';
  tree: ArchiveTree;
  version: string;
  slug: string;
  license: {
    name: string;
    version: string;
  };
  publish_date: string;
  authors: Array<{
    value: {
      name: string;
    }
  }>;
}

export interface Page {
  id: string;
  shortId: string;
  title: string;
  version: string;
}

export interface ArchiveTreeSection {
  id: string;
  shortId: string;
  title: string;
}

export interface LinkedArchiveTree extends ArchiveTree {
  parent?: LinkedArchiveTree;
}

export interface LinkedArchiveTreeSection extends ArchiveTreeSection {
  parent: LinkedArchiveTree;
}

export interface ArchiveTree extends ArchiveTreeSection {
  contents: Array<ArchiveTree | ArchiveTreeSection>;
}

export interface ArchiveBook {
  id: string;
  shortId: string;
  title: string;
  tree: ArchiveTree;
  version: string;
  license: {
    name: string;
    version: string;
  };
}

export interface ArchivePage {
  id: string;
  shortId: string;
  content: string;
  version: string;
  title: string;
}

export type ArchiveContent = ArchivePage | ArchiveBook;
