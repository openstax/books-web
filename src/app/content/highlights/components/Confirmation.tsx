import Color from 'color';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components/macro';
import Button, { ButtonGroup } from '../../../components/Button';
import { labelStyle } from '../../../components/Typography';
import theme from '../../../theme';

// tslint:disable-next-line:variable-name
export const Overlay = styled.div`
  background: ${Color(theme.color.black).alpha(0.90).string()};
  transition: background 200ms;
  position: absolute;
  display: flex;
  flex-direction: column;
  padding: 0.8rem 1.6rem;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  overflow: hidden;

  label {
    ${labelStyle}
    color: ${theme.color.text.white};
  }
`;

interface Props {
  message: string;
  onConfirm: () => void;
  onCancel: () => void;
  always: () => void;
}

// tslint:disable-next-line:variable-name
const Confirmation = ({message, always, onCancel, onConfirm}: Props) => <Overlay>
  <FormattedMessage id={message}>
    {(msg: Element | string) => <label>{msg}</label>}
  </FormattedMessage>
  <ButtonGroup>
    <Button size='small' variant='primary' onClick={(e: React.FormEvent) => {
      e.preventDefault();
      onConfirm();
      always();
    }}>Save</Button>
    <Button size='small' onClick={(e: React.FormEvent) => {
      e.preventDefault();
      onCancel();
      always();
    }}>Cancel</Button>
  </ButtonGroup>
</Overlay>;

export default Confirmation;
