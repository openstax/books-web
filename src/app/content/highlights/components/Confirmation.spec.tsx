import React from 'react';
import { createIntl, RawIntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import en from '../../../../app/messages/en/index';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { makeFindByTestId } from '../../../../test/reactutils';
import * as Services from '../../../context/Services';
import { Store } from '../../../types';
import Confirmation from './Confirmation';

// this is a hack because useEffect is currently not called
// when using jsdom? https://github.com/facebook/react/issues/14050
// seems to work better in react-test-renderer but
// i need the ref here
jest.mock('react', () => {
  const react = (jest as any).requireActual('react');
  return { ...react, useEffect: react.useLayoutEffect };
});

describe('Confirmation', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;

  const intl = createIntl({
    locale: 'en',
    messages: {
      ...en,
      'i18n:confirm': 'confirm',
      'i18n:message': 'message',
    },
  });

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
  });

  it('matches snapshot no selection', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            data-analytics-region='region'
            confirmMessage='confirm'
            onConfirm={() => null}
            onCancel={() => null}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('prevents default when clicking confirm button', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            confirmMessage='confirm'
            onCancel={() => null}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    const button = findByTestId('confirm');

    const preventDefault = jest.fn();
    button.props.onClick({preventDefault});

    expect(preventDefault).toHaveBeenCalled();
  });

  it('prevents default when clicking cancel button', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            confirmMessage='confirm'
            onCancel={() => null}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    const button = findByTestId('cancel');

    const preventDefault = jest.fn();
    button.props.onClick({preventDefault});

    expect(preventDefault).toHaveBeenCalled();
  });

  it('doesn\'t prevent default when clicking confirm link', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            confirmLink='/asdf'
            message='message'
            confirmMessage='confirm'
            onCancel={() => null}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    const button = findByTestId('confirm');

    const preventDefault = jest.fn();
    button.props.onClick({preventDefault});

    expect(preventDefault).not.toHaveBeenCalled();
  });

  it('calls onConfirm', () => {
    const onConfirm = jest.fn();
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            confirmMessage='confirm'
            onConfirm={onConfirm}
            onCancel={() => null}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    const button = findByTestId('confirm');
    button.props.onClick({preventDefault: jest.fn()});

    expect(onConfirm).toHaveBeenCalled();
  });

  it('calls onCancel', () => {
    const onCancel = jest.fn();
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            confirmMessage='confirm'
            onConfirm={() => null}
            onCancel={onCancel}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    const button = findByTestId('cancel');
    button.props.onClick({preventDefault: jest.fn()});

    expect(onCancel).toHaveBeenCalled();
  });

  it('calls always', () => {
    const always = jest.fn();
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <RawIntlProvider value={intl}>
          <Confirmation
            message='message'
            confirmMessage='confirm'
            onCancel={() => null}
            always={always}
          />
        </RawIntlProvider>
      </Services.Provider>
    </Provider>);

    const findByTestId = makeFindByTestId(component.root);
    findByTestId('confirm').props.onClick({preventDefault: jest.fn()});
    findByTestId('cancel').props.onClick({preventDefault: jest.fn()});

    expect(always).toHaveBeenCalledTimes(2);
  });
});
