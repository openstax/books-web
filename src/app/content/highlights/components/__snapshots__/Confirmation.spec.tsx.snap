// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirmation matches snapshot no selection 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  font-size: 1.4rem;
  height: 3rem;
  padding: 0 2rem;
  color: #fff;
  background-color: #f36b32;
  font-weight: bold;
  border: none;
}

.c2:hover {
  background-color: #e96128;
}

.c2:active {
  background-color: #df571e;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  font-size: 1.4rem;
  height: 3rem;
  padding: 0 2rem;
  color: #5e6062;
  background-color: #fff;
  border: 1px solid #d5d5d5;
  font-weight: regular;
}

.c3:hover {
  background-color: #fafafa;
}

.c3:active {
  background-color: #e5e5e5;
}

.c1 {
  display: grid;
  overflow: visible;
  grid-auto-flow: column;
  grid-gap: 1rem;
}

.c0 {
  background: rgba(0,0,0,0.9);
  outline: none;
  border-radius: 0.4rem;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.14),0 2px 2px 0 rgba(0,0,0,0.12),0 1px 3px 0 rgba(0,0,0,0.2);
  -webkit-transition: background 200ms;
  transition: background 200ms;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.8rem 1.6rem;
  top: 0;
  right: 0;
  left: 0;
  overflow: visible;
  min-height: 100%;
}

.c0 label {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  line-height: 1.6rem;
  font-weight: normal;
  color: #fff;
  margin-bottom: 0.8rem;
}

<div
  className="c0"
  tabIndex={-1}
>
  <label>
    message
  </label>
  <div
    className="c1"
  >
    <button
      className="c2"
      data-testid="confirm"
      onClick={[Function]}
      size="small"
    >
      confirm
    </button>
    <button
      className="c3"
      data-testid="cancel"
      onClick={[Function]}
      size="small"
    >
      Cancel
    </button>
  </div>
</div>
`;
